File: app/src/main/java/com/websarva/wings/android/zuboradiary/ui/diary/diaryshow/DiaryShowViewModel.java
package com.websarva.wings.android.zuboradiary.ui.diary.diaryshow;

import android.net.Uri;

import androidx.lifecycle.LiveData;

import com.websarva.wings.android.zuboradiary.data.AppMessage;
import com.websarva.wings.android.zuboradiary.data.database.DiaryEntity;
import com.websarva.wings.android.zuboradiary.data.database.DiaryRepository;
import com.websarva.wings.android.zuboradiary.data.diary.Condition;
import com.websarva.wings.android.zuboradiary.data.diary.ItemNumber;
import com.websarva.wings.android.zuboradiary.data.diary.Weather;
import com.websarva.wings.android.zuboradiary.ui.BaseViewModel;
import com.websarva.wings.android.zuboradiary.ui.diary.DiaryLiveData;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.Objects;
import java.util.concurrent.CancellationException;
import java.util.concurrent.ExecutionException;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;

@HiltViewModel
public class DiaryShowViewModel extends BaseViewModel {

    private final DiaryRepository diaryRepository;

    // 日記データ関係
    private final DiaryLiveData diaryLiveData;

    @Inject
    public DiaryShowViewModel(DiaryRepository diaryRepository) {
        this.diaryRepository = diaryRepository;
        diaryLiveData = new DiaryLiveData();
        initialize();
    }

    @Override
    public void initialize() {
        super.initialize();
        diaryLiveData.initialize();
    }

    public boolean existsSavedDiary(LocalDate date) {
        Objects.requireNonNull(date);

        try {
            return diaryRepository.existsDiary(date).get();
        } catch (ExecutionException | InterruptedException e) {
            addAppMessage(AppMessage.DIARY_INFO_LOADING_ERROR);
            return false;
        }
    }

    public void loadSavedDiary(LocalDate date) {
        Objects.requireNonNull(date);

        try {
            DiaryEntity diaryEntity = diaryRepository.loadDiary(date).get();
            diaryLiveData.update(diaryEntity);
        } catch (Exception e) {
            addAppMessage(AppMessage.DIARY_LOADING_ERROR);
        }
    }

    boolean deleteDiary() {
        LocalDate deleteDate = diaryLiveData.getDateMutableLiveData().getValue();
        Objects.requireNonNull(deleteDate);

        Integer result;
        try {
            result = diaryRepository.deleteDiary(deleteDate).get();
        } catch (CancellationException | ExecutionException | InterruptedException  e) {
            addAppMessage(AppMessage.DIARY_DELETE_ERROR);
            return false;
        }

        // 削除件数 = 1が正常
        if (result != 1) {
            addAppMessage(AppMessage.DIARY_DELETE_ERROR);
            return false;
        }

        return true;
    }

    // MEMO:存在しないことを確認したいため下記メソッドを否定的処理とする
    boolean checkSavedPicturePathDoesNotExist(Uri uri) {
        Objects.requireNonNull(uri);

        try {
            return !diaryRepository.existsPicturePath(uri).get();
        } catch (ExecutionException | InterruptedException e) {
            addAppMessage(AppMessage.DIARY_LOADING_ERROR);
            return false;
        }
    }

    // LiveDataGetter
    public LiveData<LocalDate> getDateLiveData() {
        return diaryLiveData.getDateMutableLiveData();
    }

    public LiveData<Weather> getWeather1LiveData() {
        return diaryLiveData.getWeather1MutableLiveData();
    }

    public LiveData<Weather> getWeather2LiveData() {
        return diaryLiveData.getWeather2MutableLiveData();
    }

    public LiveData<Condition> getConditionLiveData() {
        return diaryLiveData.getConditionMutableLiveData();
    }

    public LiveData<String> getTitleLiveData() {
        return diaryLiveData.getTitleMutableLiveData();
    }

    public LiveData<Integer> getNumVisibleItemsLiveData() {
        return diaryLiveData.getNumVisibleItemsMutableLiveData();
    }

    public LiveData<String> getItem1TitleLiveData() {
        return diaryLiveData.getItemLiveData(new ItemNumber(1)).getTitleMutableLiveData();
    }

    public LiveData<String> getItem2TitleLiveData() {
        return diaryLiveData.getItemLiveData(new ItemNumber(2)).getTitleMutableLiveData();
    }

    public LiveData<String> getItem3TitleLiveData() {
        return diaryLiveData.getItemLiveData(new ItemNumber(3)).getTitleMutableLiveData();
    }

    public LiveData<String> getItem4TitleLiveData() {
        return diaryLiveData.getItemLiveData(new ItemNumber(4)).getTitleMutableLiveData();
    }

    public LiveData<String> getItem5TitleLiveData() {
        return diaryLiveData.getItemLiveData(new ItemNumber(5)).getTitleMutableLiveData();
    }

    public LiveData<String> getItem1CommentLiveData() {
        return diaryLiveData.getItemLiveData(new ItemNumber(1)).getCommentMutableLiveData();
    }

    public LiveData<String> getItem2CommentLiveData() {
        return diaryLiveData.getItemLiveData(new ItemNumber(2)).getCommentMutableLiveData();
    }

    public LiveData<String> getItem3CommentLiveData() {
        return diaryLiveData.getItemLiveData(new ItemNumber(3)).getCommentMutableLiveData();
    }

    public LiveData<String> getItem4CommentLiveData() {
        return diaryLiveData.getItemLiveData(new ItemNumber(4)).getCommentMutableLiveData();
    }

    public LiveData<String> getItem5CommentLiveData() {
        return diaryLiveData.getItemLiveData(new ItemNumber(5)).getCommentMutableLiveData();
    }

    public LiveData<Uri> getPicturePathLiveData() {
        return diaryLiveData.getPicturePathMutableLiveData();
    }

    public LiveData<LocalDateTime> getLogLiveData() {
        return diaryLiveData.getLogMutableLiveData();
    }
}

File: app/src/main/java/com/websarva/wings/android/zuboradiary/ui/diary/diaryshow/DiaryShowViewModel.kt
package com.websarva.wings.android.zuboradiary.ui.diary.diaryshow

import android.net.Uri
import android.util.Log
import com.websarva.wings.android.zuboradiary.data.database.DiaryRepository
import com.websarva.wings.android.zuboradiary.data.diary.ItemNumber
import com.websarva.wings.android.zuboradiary.createLogTag
import com.websarva.wings.android.zuboradiary.data.DiaryShowAppMessage
import com.websarva.wings.android.zuboradiary.ui.BaseViewModel
import com.websarva.wings.android.zuboradiary.ui.DiaryShowPendingDialog
import com.websarva.wings.android.zuboradiary.ui.requireValue
import com.websarva.wings.android.zuboradiary.ui.diary.DiaryStateFlow
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.asStateFlow
import java.time.LocalDate
import javax.inject.Inject

@HiltViewModel
internal class DiaryShowViewModel @Inject constructor(private val diaryRepository: DiaryRepository) :
    BaseViewModel() {

    private val logTag = createLogTag()

    // 日記データ関係
    private val diaryStateFlow = DiaryStateFlow()
    val date
        get() = diaryStateFlow.date.asStateFlow()
    val weather1
        get() = diaryStateFlow.weather1.asStateFlow()
    val weather2
        get() = diaryStateFlow.weather2.asStateFlow()
    val condition
        get() = diaryStateFlow.condition.asStateFlow()
    val title
        get() = diaryStateFlow.title.asStateFlow()
    val numVisibleItems
        get() = diaryStateFlow.numVisibleItems.asStateFlow()
    val item1Title
        get() = diaryStateFlow.getItemStateFlow(ItemNumber(1)).title.asStateFlow()
    val item2Title
        get() = diaryStateFlow.getItemStateFlow(ItemNumber(2)).title.asStateFlow()
    val item3Title
        get() = diaryStateFlow.getItemStateFlow(ItemNumber(3)).title.asStateFlow()
    val item4Title
        get() = diaryStateFlow.getItemStateFlow(ItemNumber(4)).title.asStateFlow()
    val item5Title
        get() = diaryStateFlow.getItemStateFlow(ItemNumber(5)).title.asStateFlow()
    val item1Comment
        get() = diaryStateFlow.getItemStateFlow(ItemNumber(1)).comment.asStateFlow()
    val item2Comment
        get() = diaryStateFlow.getItemStateFlow(ItemNumber(2)).comment.asStateFlow()
    val item3Comment
        get() = diaryStateFlow.getItemStateFlow(ItemNumber(3)).comment.asStateFlow()
    val item4Comment
        get() = diaryStateFlow.getItemStateFlow(ItemNumber(4)).comment.asStateFlow()
    val item5Comment
        get() = diaryStateFlow.getItemStateFlow(ItemNumber(5)).comment.asStateFlow()
    val picturePath
        get() = diaryStateFlow.picturePath.asStateFlow()
    val log
        get() = diaryStateFlow.log.asStateFlow()

    override fun initialize() {
        super.initialize()
        diaryStateFlow.initialize()
    }

    suspend fun loadSavedDiary(date: LocalDate, ignoreAppMessage: Boolean = false): Boolean {
        val logMsg = "日記読込"
        Log.i(logTag, "${logMsg}_開始")

        try {
            val diaryEntity = diaryRepository.loadDiary(date) ?: throw IllegalArgumentException()
            diaryStateFlow.update(diaryEntity)
        } catch (e: Exception) {
            Log.e(logTag, "${logMsg}_失敗", e)
            if (!ignoreAppMessage) {
                addAppMessage(DiaryShowAppMessage.DiaryLoadingFailure)
            }
            return false
        }

        Log.i(logTag, "${logMsg}_完了")
        return true
    }

    suspend fun deleteDiary(): Boolean {
        val logMsg = "日記削除"
        Log.i(logTag, "${logMsg}_開始")

        val deleteDate = diaryStateFlow.date.requireValue()
        try {
            diaryRepository.deleteDiary(deleteDate)
        } catch (e: Exception) {
            Log.e(logTag, "${logMsg}_失敗", e)
            addAppMessage(DiaryShowAppMessage.DiaryDeleteFailure)
            return false
        }

        Log.i(logTag, "${logMsg}_完了")
        return true
    }

    // MEMO:存在しないことを確認したいため下記メソッドを否定的処理とする
    suspend fun checkSavedPicturePathDoesNotExist(uri: Uri): Boolean? {
        try {
            return !diaryRepository.existsPicturePath(uri)
        } catch (e: Exception) {
            Log.e(logTag, "端末写真URI使用状況確認_失敗", e)
            addAppMessage(DiaryShowAppMessage.DiaryLoadingFailure)
            return null
        }
    }

    // 表示保留中Dialog追加
    // MEMO:引数の型をサブクラスに制限
    fun addPendingDialogList(pendingDialog: DiaryShowPendingDialog) {
        super.addPendingDialogList(pendingDialog)
    }
}
